control-plane-operator/controllers/hostedcontrolplane/hostedcontrolplane_controller.go

```diff
@@ -180,7 +184,12 @@ func (r *HostedControlPlaneReconciler) Reconcile(ctx context.Context, req ctrl.R
 			Type:               string(hyperv1.ValidConfiguration),
 			ObservedGeneration: hostedControlPlane.Generation,
 		}
-		if err := config.ValidateGlobalConfig(ctx, hostedControlPlane); err != nil {
+		// TODO: Should there already be a config validation in the hostedcluster controller?
+		if err := r.validateConfigAndClusterCapabilities(hostedControlPlane); err != nil {
```

Comment:
I'm confused why was this moved from hosted cluster controller and validating the HostedCluster? https://github.com/openshift/hypershift/pull/604#discussion_r737523546
That's the first entry point that needs to be validated, can we please put it back? We can keep it here as well.

Reply:
> I'm confused why was this moved from hosted cluster controller and validating the HostedCluster? [#604 (comment)](https://github.com/openshift/hypershift/pull/604#discussion_r737523546)

Yes this is all a little bit confusing. I agree that all validation should happen already before too, but it seems like all validation is right now exclusively done in the hostedcontrolplane controller and not the hostedcluster controller.

This already led to all the confusion before. The InvalidClusterConfig condition in the hostedcluster controller is bubbled up from the InvalidConfig condition on the hostedcontrolplane controller.

Due to this I thought it might be best to keep all config validation in one place right now and add a TODO.

Reply:
>The InvalidClusterConfig condition in the hostedcluster controller is bubbled up from the InvalidConfig condition on the hostedcontrolplane controller.

Can you point me to where that's happening?

>but it seems like all validation is right now exclusively done in the hostedcontrolplane controller and not the hostedcluster controller.

What do you mean exactly by all validation? The main validation happening in the hostedcontrolplane is for global config which is intrinsic to the control plane and it's control plane version dependent so it makes sense for it to be there.

This is different in the sense that is not tied to the hcp it's invalid user input for the HC and it should be computed and signalled there before anywhere else.

Reply:
> > The InvalidClusterConfig condition in the hostedcluster controller is bubbled up from the InvalidConfig condition on the hostedcontrolplane controller.
> 
> Can you point me to where that's happening?

Here: https://github.com/openshift/hypershift/blob/main/hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go#L294

this is as far as I could see the only place in the code which would set this condition to false. It bubbles up the condition from the hostedcontrolplane CR.

> 
> > but it seems like all validation is right now exclusively done in the hostedcontrolplane controller and not the hostedcluster controller.
> 
> What do you mean exactly by all validation? The main validation happening in the hostedcontrolplane is for global config which is intrinsic to the control plane and it's control plane version dependent so it makes sense for it to be there.

Ok, so you mean that you intentionally do not validate this section before.

> 
> This is different in the sense that is not tied to the hcp it's invalid user input for the HC and it should be computed and signalled there before anywhere else.

I am not sure. I guess the same could apply like above. That it may be version dependent. If a new fields get added which the underlying hostedcontrolplane controller will ignore because it does not even know about it yet.

Adding it to both places works for me too, but I guess the two controllers would would then potentially try to occupy the same condition (for the same reason but with different code paths).

Reply:
>I am not sure. I guess the same could apply like above. That it may be version dependent. If a new fields get added which the underlying hostedcontrolplane controller will ignore because it does not even know about it yet.

What I mean is Hypershift owns the lifecycle of HC and HCP APIs. Now, the HCP operator implementation it's tied and might differ for each OCP release payload and the global config being validated there it's config for OCP specific version operators. We reserve the right to change the validation implementation for different ocp releases if we want/need to.
Now for a HostedCluster it is always invalid that you set a Route strategy if that's not supported by the env. No matter which version of the HCP/OCP you are wanting to run. And if we want to change that behaviour or semantic in the future we can independently from OCP since that's our API and semantics.

Reply:
Ok makes sense. So from that perspective where I had the code before it was more correct? Do you want me to introduce another condition or should I restore my previous if/else logic where I would give the cluster wide validation precedence?

Reply:
I think my preference would be to have a new condition that is a semantic for HC invalid input and possibly rename the one that signal the HCP global config to ValidHCPConfiguration or similar https://github.com/openshift/hypershift/blob/main/hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go#L306

But first I'd like to make sure @ironcladlou and @csrwng (who originally implemented the global config) agree with the above and https://github.com/openshift/hypershift/pull/604#discussion_r740899781 before proceeding.

Reply:
Thanks for the offline discussion. I think I rearranged the code as we discussed. There is one open point: In case that somehow a controlplaneconfig which wants unavailable capabilities gets propagated, the hostedcontrolplane controller is performing a capabilities check too an reports it in its own condition and does not reconcile like with any other invalid config. Otherwise one would only see in the controller logs that it failed on e.g. creating routes. I can remove that too or just leave it. 

PTAL.